pkg/gateway/model_build_access_log_subscription.go

```diff
@@ -57,26 +58,29 @@ type accessLogSubscriptionModelBuildTask struct {
 }
 
 func (t *accessLogSubscriptionModelBuildTask) run(ctx context.Context) error {
+	var eventType = core.CreateEvent
+	if t.accessLogPolicy.DeletionTimestamp != nil {
+		eventType = core.DeleteEvent
+	} else if _, ok := t.accessLogPolicy.Annotations[anv1alpha1.AccessLogSubscriptionAnnotationKey]; ok {
+		eventType = core.UpdateEvent
+	}
+
 	sourceType := model.ServiceSourceType
 	if t.accessLogPolicy.Spec.TargetRef.Kind == "Gateway" {
 		sourceType = model.ServiceNetworkSourceType
 	}
 
 	sourceName, err := utils.TargetRefToLatticeResourceName(t.accessLogPolicy.Spec.TargetRef, t.accessLogPolicy.Namespace)
-	if err != nil {
+	if err != nil && eventType != core.DeleteEvent {
 		return err
 	}
 
 	destinationArn := t.accessLogPolicy.Spec.DestinationArn
 	if destinationArn == nil {
-		return fmt.Errorf("access log policy's destinationArn cannot be nil")
-	}
-
-	var eventType = core.CreateEvent
-	if t.accessLogPolicy.DeletionTimestamp != nil {
-		eventType = core.DeleteEvent
-	} else if _, ok := t.accessLogPolicy.Annotations[anv1alpha1.AccessLogSubscriptionAnnotationKey]; ok {
-		eventType = core.UpdateEvent
+		if eventType != core.DeleteEvent {
+			return fmt.Errorf("access log policy's destinationArn cannot be nil")
+		}
+		destinationArn = aws.String("")
```

Comment:
I dont have good suggestion yet, but it bothers me that we do conditional code branching on every step - reconciler, model, synthezier. Every time we do something that checks if it's delete/update/create, makes it harder to reason about how it actually works.

I'm thinking that create/update/delete branching should happen at Reconcile function. The rest of the code should not have these. It might create code duplicates, but those can be written into smaller functions. Right now I have to look at every stage and assemble little piece of what is delete logic and build some mental map of it.

It's not a blocker, but I believe we come back to this code soon.

Reply:
Agreed, I think this is going to be related to #439 and #457